azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java

```diff
@@ -5,34 +5,176 @@
+    private Map<String, TokenCredential> tenantToCredential = new HashMap<>();
 
     public Account logout() {
-        this.authenticated = false;
+        this.entity = null;
+        this.credentialBuilder = null;
+        this.tenantToCredential = new HashMap<>();
         return this;
     }
 
-    /**
-     * @return the credential for specified subscription
-     */
-    public AzureTokenCredentials getCredential(String subscriptionId) {
+    public AzureEnvironment getEnvironment() {
+        return entity.getEnvironment();
+    }
+
+    public void initialize() {
+        Objects.requireNonNull(entity, "Cannot initialize from null account entity.");
+        AzureEnvironment env = MoreObjects.firstNonNull(getEnvironment(), AzureEnvironment.AZURE);
+        AzureProfile azureProfile = new AzureProfile(env);
+        Set<String> validTenantIds = new HashSet<>();
+        if (this.entity.getTenantIds() == null) {
+            this.entity.setTenantIds(AzureResourceManager.authenticate(this.credentialBuilder.provideCredentialCommon()
+                    , azureProfile).tenants().list().stream().map(Tenant::tenantId).collect(Collectors.toList()));
+
+        }
+
+        Map<String, SubscriptionEntity> subscriptionMap = new HashMap<>();
+        this.entity.getTenantIds().forEach(tenantId -> {
+            try {
+                List<SubscriptionEntity> subscriptionsOnTenant =
+                        AzureResourceManager.authenticate(this.credentialBuilder.provideCredentialForTenant(tenantId), azureProfile).subscriptions().list()
+                                .mapPage(s -> this.toSubscriptionEntity(entity.getEnvironment(), tenantId, s)).stream().collect(Collectors.toList());
+
+                for (SubscriptionEntity s : subscriptionsOnTenant) {
+                    String key = StringUtils.lowerCase(s.getId());
+                    if (subscriptionMap.putIfAbsent(key, s) != null) {
+                        validTenantIds.add(tenantId);
+                    }
+                }
+            } catch (Exception ex) {
+                // ignore AuthenticationException since on some tenants, it doesn't allow list subscriptions
+                if (!(ExceptionUtils.getRootCause(ex) instanceof AuthenticationException)) {
+                    this.entity.setAuthenticated(false);
+                    Account.this.entity.setError(ex);
+                }
+
+            }
+        });
+
+        // since some tenants doesn't have subscriptions, reduce the tenants
+        this.entity.setTenantIds(new ArrayList<>(validTenantIds));
+        this.entity.setSubscriptions(new ArrayList<>(subscriptionMap.values()));
+        if (subscriptionMap.isEmpty()) {
+            this.entity.setAuthenticated(false);
+        }
+    }
+
+    public void selectSubscriptions(List<String> subscriptionIds) {
+        // select subscriptions
+        if (CollectionUtils.isNotEmpty(subscriptionIds) && CollectionUtils.isNotEmpty(this.entity.getSubscriptions())) {
```

Comment:
Do we need to set all subscriptions unselected to clear previous selection status?

Reply:
selectSubscriptions should be private can be only called once and cannot be called again.